src/integrationTest/java/org/opensearch/test/framework/cluster/LocalOpenSearchCluster.java

```diff
@@ -0,0 +1,491 @@
+        this.clusterName = clusterName;
+        this.clusterManager = clusterManager;
+        this.nodeSettingsSupplier = nodeSettingsSupplier;
+        this.additionalPlugins = additionalPlugins;
+        this.testCertificates = testCertificates;
+        try {
+			this.clusterHomeDir = Files.createTempDirectory("local_cluster_" + clusterName).toFile();
+		} catch (IOException e) {
+			throw new IllegalStateException(e);
+		}        
+    }
+
+    private List<Node> getNodesByType(NodeType nodeType) {
+        return nodes.stream()
+            .filter(currentNode -> currentNode.hasAssignedType(nodeType))
+            .collect(Collectors.toList());
+    }
+
+    private long countNodesByType(NodeType nodeType) {
+        return getNodesByType(nodeType).stream().count();
+    }
+
+    public void start() throws Exception {
+        log.info("Starting {}", clusterName);
+
+        int masterNodeCount = clusterManager.getMasterNodes();
+        int nonMasterNodeCount = clusterManager.getDataNodes() + clusterManager.getClientNodes();
+
+        SortedSet<Integer> masterNodeTransportPorts = PortAllocator.TCP.allocate(clusterName, Math.max(masterNodeCount, 4), 5000 + 42 * 1000 + 300);
+        SortedSet<Integer> masterNodeHttpPorts = PortAllocator.TCP.allocate(clusterName, masterNodeCount, 5000 + 42 * 1000 + 200);
+
+        this.seedHosts = toHostList(masterNodeTransportPorts);
+        this.initialMasterHosts = toHostList(masterNodeTransportPorts.stream().limit(masterNodeCount).collect(Collectors.toSet()));
+
+        started = true;
+
+        CompletableFuture<Void> masterNodeFuture = startNodes(
+            clusterManager.getMasterNodeSettings(), masterNodeTransportPorts,
+                masterNodeHttpPorts);
+
+        SortedSet<Integer> nonMasterNodeTransportPorts = PortAllocator.TCP.allocate(clusterName, nonMasterNodeCount, 5000 + 42 * 1000 + 310);
+        SortedSet<Integer> nonMasterNodeHttpPorts = PortAllocator.TCP.allocate(clusterName, nonMasterNodeCount, 5000 + 42 * 1000 + 210);
+
+        CompletableFuture<Void> nonMasterNodeFuture = startNodes(
+            clusterManager.getNonMasterNodeSettings(), nonMasterNodeTransportPorts,
+                nonMasterNodeHttpPorts);
+
+        CompletableFuture.allOf(masterNodeFuture, nonMasterNodeFuture).join();
+
+        if (isNodeFailedWithPortCollision()) {
+            log.info("Detected port collision for master node. Retrying.");
+
+            retry();
+            return;
+        }
+
+        log.info("Startup finished. Waiting for GREEN");
+
+        waitForCluster(ClusterHealthStatus.GREEN, TimeValue.timeValueSeconds(10), nodes.size());
+
+        log.info("Started: {}", this);
+
+    }
+
+    public String getClusterName() {
+        return clusterName;
+    }
+
+    public boolean isStarted() {
+        return started;
+    }
+
+    public void stop() {
+        List<CompletableFuture<Boolean>> stopFutures = new ArrayList<>();
+        for (Node node : getNodesByType(CLIENT)) {
```

Comment:
Does this need to happen by node type, since there is no delay these are effectively being issued asynchronously all at once, it would be easier to have a single stream map'ing to completable futures they are waited on

Reply:
done